core/node/api_server/src/web3/namespaces/en.rs

```diff
@@ -36,6 +81,23 @@ impl EnNamespace {
         )))
     }
 
+    #[tracing::instrument(skip(self))]
+    pub async fn simulated_l1_status_impl(&self) -> Result<en::SimulatedL1Status, Web3Error> {
+        let mut conn = self.state.acquire_connection().await?;
+        let next_batch_to_commit = match conn
+            .consensus_dal()
+            .get_last_batch_certificate_number()
+            .await
+            .context("get_last_batch_certificate_number()")?
+        {
+            Some(n) => to_l1_batch_number(n + 1)?,
+            None => self.first_batch_to_commit,
+        };
```

Comment:
why do we have this state at all? Why not apply the logic of `EnNamespace::new` on every API call? I'm probably missing something since now it feels like we have some extra in-memory state (`first_batch_to_commit`) that is not really needed.

Reply:
we need to return consistent responses when the batch certificates table is empty

Reply:
Can the table get pruned while the API is running, and then fall back to a very old `first_batch_to_commit` that has been surpassed since, regressing the value? 

In other words, can this happen?
1. Start the node, create an `EnNamespace` instance, assign `first_batch_to_commit = 100`
2. Batches are created, QCs are saved, saw we go up to batch number 500
3. DB gets pruned, including the batch QCs
4. Next request sees the 100 again?

Reply:
What do you mean exactly by inconsistent results?

Looking at the tests, my guess is that because `first_batch_to_commit` relies on `get_sealed_l1_batch_number` before the first batch QC is created, if we called it again without a QC it would move on as soon as there is a new L1 batch.

The problem I see is that if the main node is restarted, it will move on anyway, because it forgets the previous `first_batch_to_commit` value. Currently that seems fine; if we haven't started creating batches yet, we might as well keep returning the latest value until a QC is formed, then go consistently from there. 

I'm not sure how it works with pruning and snapshotting though. Ideally if there _were_ batch QCs before, then it could tell what the latest was before the snapshot, and continue without gaps. However, it can only ask for a QC of something for which it can verify the hash as well, so it might just have to be whatever the next L1 batch record is going to be.

Reply:
On second thought, I can see why we want to stick to some value while the node is running: I expect that I will also initialise `BatchVotes::min_batch_number` to something, and ideally it should agree with what is returned here. If it doesn't agree, then attesters might be voting on something that the main node ignores, although hopefully just temporarily. A disagreement can stem from a race condition: between the initialisation of `EnNamespace` and `BatchVotes` an extra L1 batch might be created. 

The need to agree would be another reason to delegate this whole query fallback logic to the DAL, so that it's available to both components. Maybe it can be cached there?

Reply:
A bad race condition would be this:
1. `EnNamespace` is created, there are no batch QC yet, and `first_batch_to_commit` is 100
2. A batch is created on the main node
3. `BatchVotes` is created on the main node and the `min_batch_number` is set to 101
4. External Nodes query this API and get 100, they cast vote on that, but the main node ignores it because it wants at least 101

If the API always returned the latest, then `BatchVotes` would certainly accept votes on 102, 103, etc, even if its minimum is set to 101. 

Otherwise I'd say this number could be in an input for this component, and then the compositor would have to make sure it's correctly initialised to the same value as the other.

Reply:
> we need to return consistent responses when the batch certificates table is empty

Reply:
> we need to return consistent responses when the batch certificates table is empty

Makes sense, but we lose it if the node is restarted anyway right? So we cannot rely on the fact that it's consistent. Do we want it because of

> On second thought, I can see why we want to stick to some value while the node is running: I expect that I will also initialise BatchVotes::min_batch_number to something, and ideally it should agree with what is returned here. If it doesn't agree, then attesters might be voting on something that the main node ignores, although hopefully just temporarily. A disagreement can stem from a race condition: between the initialisation of EnNamespace and BatchVotes an extra L1 batch might be created.

If so, we probably need a single source of truth for the both.

Reply:
There is no pruning on the main node, my implementation is misleading indeed. It also doesn't matter what this rpc serves on the EN (in fact, it would be better to disable this rpc on the EN, but I guess we don't care).

Reply:
By consistent responses, I mean that on main node startup (with empty l1_batches_consensus), the batch to sign shouldn't be a moving target, i.e. we shouldn't just return the last_sealed_block whatever it is.

Reply:
I think it would be okay to return whatever the `last_sealed_block` is. Like you pointed out, whether it's +1 or +5 doesn't matter, and after a restart it will move anyway. In fact if it moves, and the `BatchVotes::set_min_batch_number` is updated with it (like the ENs will have to do via polling the API), then we won't get stuck and eventually a QC will be formed. Once a QC is formed, it will stop moving on its own.

Reply:
I've removed querying the pruning_info, ptal

Reply:
@aakoshh I don't want a race condition between new batches sealed and bootstraping the attestation.

Reply:
> By consistent responses, I mean that on main node startup (with empty l1_batches_consensus), the batch to sign shouldn't be a moving target, i.e. we shouldn't just return the last_sealed_block whatever it is.

Why? Won't it be the case anyway if the node restarts?

Reply:
> I don't want a race condition between new batches sealed and bootstraping the attestation.

Sorry, I didn't want to give the impression that I would like a race condition to exist. I can imagine that the node builder makes this impossible even, with components being initialised first, and during this time I assume it won't create new L1 batches. I just wanted to think through what could go wrong if there was one, without assuming anything about the rest of the system.

My follow up for this PR would be to add a loop that polls what the next batch to vote on should be, and pass it to `BatchVotes::set_min_number`; the source of the data would be the DAL on the main node, and the API on the external nodes. Which is why I thought the DAL could be a good home for this logic, potentially with a cache for consistency, but even easier if it returns the last sealed batch (+1 if you will) until there is a batch QC.

That way I don't think there can be any race conditions, just a bit of eventual consistency between the MN and the ENs: we can initialise the `BatchVotes` with whatever the DAL/API returns to prevent votes from the irrelevant past, then keep increasing the floor and hope that at some point the MN will keep it stable long enough for API queries to refresh votes to arrive within an L1 window; from that point it will be tied to the last batch QC and stay stable.

Reply:
> Why? Won't it be the case anyway if the node restarts?

We don't care about restarts, because the collecting votes happens during single main node execution.

Reply:
> My follow up for this PR would be to add a loop that polls what the next batch to vote on should be, and pass it to BatchVotes::set_min_number

sg, except that (if possible) the main node should rather its own api (for symmetry)

Reply:
Just to clarify what querying the main node API from the main node itself would mean in terms of deployment:

As I understand the API is running in a different container from the main node, so there will be some new startup dependency between them if I want to initialise it the vote registry. Not sure how it worked with local testing, I guess the component list must have included both consensus and web3. 

I would have thought it would be better to move the logic into the DAL layer and make both web3 and consensus depend on them, keep web3 thin. I see now that even if there was caching of the first in the DAL layer it wouldn’t help keep things consistent because they would be in different containers. 

I still think it could be made stateless to avoid all this head scratching. 

What if the web3 layer is load balanced and everyone sees different values based on when instances were started?

Reply:
I am also not 100% clear now API based init would work with the node builder. Presumably during wiring the API isn’t running, so the consensus layer would either have to init to 0 or do a lazy init.

Reply:
Here’s what I think (sorry if I sound like a broken record):
* For now we want the main node and its API to be the single source of truth for what to vote on. 
* After the first batch QC is inserted we don’t want gaps. 
* Before the first batch QC is inserted we want to facilitate attesters agreeing on what to vote on, so a QC is not impossible. 


If we add load balanced scalable web APIs to the picture, potentially with sticky sessions, it has the potential to prevent the first QC from being formed: instead of a single source of truth we have as many as web container instances. If attesters get stuck to querying the same instance, they might each cast their votes on something else forever. 

A stateless DB query might present a value that changes in those initial times, but it does so infrequently at the moment, and gives a minute for the first QC to form each time. 

If we move to a higher batch cadence, we would probably also introduce voting on multiple heights concurrently, in which case we could also adjust the “next” to be less aggressive in how it follows the last, and some QC will form.

Reply:
you are right, storing it in memory won't guarantee consistency because API servers are a distributed service. I'll adjust it as we discussed in the meeting.